File: vibelign/core/project_scan.py
# === ANCHOR: PROJECT_SCAN_START ===
import re
import os
from collections.abc import Generator
from pathlib import Path
from typing import TypedDict, cast

from vibelign.core.checkpoint_engine.rust_engine import scan_project_with_rust

from vibelign.core.structure_policy import (
    has_ignored_part,
    is_core_entry_file,
    is_source_file,
)


class SourceFileRecord(TypedDict):
    path: Path
    category: str | None
    imports: list[str] | None


def iter_project_files(root: Path) -> Generator[Path, None, None]:
    for path in root.rglob("*"):
        if has_ignored_part(path.parts):
            continue
        try:
            is_file = path.is_file()
        except OSError:
            continue
        if is_file:
            yield path


def iter_source_files(root: Path) -> Generator[Path, None, None]:
    for record in iter_source_file_records(root):
        yield record["path"]


def iter_source_file_records(root: Path) -> Generator[SourceFileRecord, None, None]:
    if _rust_project_scan_enabled():
        rust_records = _source_file_records_from_rust_project_scan(root)
        if rust_records is not None:
            yield from rust_records
            return
    for path in _python_source_files(root):
        yield {"path": path, "category": None, "imports": None}


def _python_source_files(root: Path) -> Generator[Path, None, None]:
    from vibelign.core.fast_tools import has_fd, find_source_files_fd

    if has_fd():
        files = find_source_files_fd(root)
        if files:
            yield from files
            return
    for path in iter_project_files(root):
        if is_source_file(path):
            yield path


def _rust_project_scan_enabled() -> bool:
    return os.environ.get("VIBELIGN_PROJECT_SCAN_RUST", "1").strip().lower() not in {
        "0",
        "false",
        "no",
    }


def _source_file_records_from_rust_project_scan(root: Path) -> list[SourceFileRecord] | None:
    report, warning = scan_project_with_rust(root)
    if warning or report is None:
        return None
    files = report.get("files")
    if not isinstance(files, list):
        return None
    records: list[SourceFileRecord] = []
    for item in cast(list[object], files):
        if not isinstance(item, dict):
            continue
        raw_item = cast(dict[object, object], item)
        path = raw_item.get("path")
        if isinstance(path, str) and path:
            rel_parts = tuple(Path(path.replace("\\", "/")).parts)
            if has_ignored_part(rel_parts):
                continue
            category = raw_item.get("category")
            raw_imports = raw_item.get("imports")
            import_items = cast(list[object], raw_imports) if isinstance(raw_imports, list) else None
            imports = [value for value in import_items if isinstance(value, str)] if import_items is not None else None
            records.append(
                {
                    "path": root / path,
                    "category": category if isinstance(category, str) else None,
                    "imports": imports,
                }
            )
    return records


def safe_read_text(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8", errors="ignore")
    except Exception:
        return ""


def line_count(path: Path) -> int:
    return len(safe_read_text(path).splitlines())


def relpath_str(root: Path, path: Path) -> str:
    try:
        return str(path.relative_to(root)).replace("\\", "/")
    except Exception:
        return str(path).replace("\\", "/")


_UI_TOKENS = ["ui", "view", "views", "window", "dialog", "widget", "screen"]
_SERVICE_TOKENS = [
    "service",
    "services",
    "api",
    "client",
    "server",
    "worker",
    "job",
    "task",
    "queue",
    "auth",
    "data",
]
_CORE_TOKENS = ["core", "engine", "patch", "anchor", "guard"]


_PY_IMPORT_RE = re.compile(
    r"^\s*(?:from\s+([.\w]+)\s+import\b|import\s+([.\w]+))",
    re.MULTILINE,
)
_JS_IMPORT_RE = re.compile(
    r"""(?:import\s+(?:[^'"`;]+?\s+from\s+)?|require\s*\(\s*)['"]([^'"]+)['"]""",
)


def extract_imports(path: Path) -> list[str]:
    """파일에서 import 대상을 raw 문자열로 추출.

    - Python: `from X import ...`, `import X` → `X`
    - JS/TS: `import ... from "X"`, `import "X"`, `require("X")` → `X`
    반환은 중복 제거된 순서 보존 리스트.
    """
    suffix = path.suffix.lower()
    text = safe_read_text(path)
    if not text:
        return []
    seen: list[str] = []
    if suffix == ".py":
        for match in _PY_IMPORT_RE.finditer(text):
            mod = match.group(1) or match.group(2)
            if mod and mod not in seen:
                seen.append(mod)
    elif suffix in {".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs"}:
        for match in _JS_IMPORT_RE.finditer(text):
            mod = match.group(1)
            if mod and mod not in seen:
                seen.append(mod)
    return seen


def classify_file(path: Path, rel: str) -> str:
    if is_core_entry_file(path):
        return "entry"
    parts = rel.replace("\\", "/").split("/")
    dir_low = "/".join(parts[:-1]).lower()
    if dir_low:
        if any(t in dir_low for t in _CORE_TOKENS):
            return "core"
        if any(t in dir_low for t in _UI_TOKENS):
            return "ui"
        if any(t in dir_low for t in _SERVICE_TOKENS):
            return "service"
    low = rel.lower()
    if any(t in low for t in _UI_TOKENS):
        return "ui"
    if any(t in low for t in _SERVICE_TOKENS):
        return "service"
    if any(t in low for t in _CORE_TOKENS):
        return "core"
    return "other"


# === ANCHOR: PROJECT_SCAN_END ===

File: vibelign-core/src/project_scan.rs
// === ANCHOR: PROJECT_SCAN_START ===
use serde::Serialize;
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

#[derive(Debug, Serialize)]
pub struct ProjectScanReport {
    pub files: Vec<ProjectScanFile>,
}

#[derive(Debug, Serialize)]
pub struct ProjectScanFile {
    pub path: String,
    pub category: String,
    pub imports: Vec<String>,
}

const IGNORED_DIRS: &[&str] = &[
    ".git",
    ".venv",
    "venv",
    "env",
    "__pycache__",
    "node_modules",
    "dist",
    "build",
    "target",
    ".next",
    ".pnpm-store",
    ".idea",
    ".vscode",
    ".pytest_cache",
    ".mypy_cache",
    ".sisyphus",
    ".trash",
    "library",
    "cloudstorage",
    "docs",
    "tests",
    ".github",
    ".vibelign",
];

const SOURCE_EXTENSIONS: &[&str] = &[
    "py", "js", "ts", "jsx", "tsx", "rs", "go", "java", "cs", "cpp", "c", "hpp", "h",
];

const CORE_ENTRY_FILES: &[&str] = &[
    "main.py",
    "app.py",
    "cli.py",
    "server.py",
    "index.js",
    "app.js",
    "main.js",
    "main.ts",
    "index.ts",
    "main.rs",
    "main.go",
    "main.cpp",
    "Program.cs",
    "vib_cli.py",
    "mcp_server.py",
];

const UI_TOKENS: &[&str] = &["ui", "view", "views", "window", "dialog", "widget", "screen"];
const SERVICE_TOKENS: &[&str] = &[
    "service", "services", "api", "client", "server", "worker", "job", "task", "queue", "auth", "data",
];
const CORE_TOKENS: &[&str] = &["core", "engine", "patch", "anchor", "guard"];

pub fn scan(root: &Path) -> Result<ProjectScanReport, String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root).into_iter().filter_entry(|entry| !is_ignored(entry.path())) {
        let entry = entry.map_err(|error| error.to_string())?;
        if !entry.file_type().is_file() || !is_source_file(entry.path()) {
            continue;
        }
        let path = relpath(root, entry.path());
        files.push(ProjectScanFile {
            category: classify_file(entry.path(), &path),
            imports: extract_imports(entry.path()),
            path,
        });
    }
    files.sort_by(|left, right| left.path.cmp(&right.path));
    Ok(ProjectScanReport { files })
}

fn is_ignored(path: &Path) -> bool {
    path.components().any(|component| {
        let part = component.as_os_str().to_string_lossy().to_lowercase();
        IGNORED_DIRS.contains(&part.as_str())
    })
}

fn is_source_file(path: &Path) -> bool {
    path.extension()
        .and_then(|extension| extension.to_str())
        .map(|extension| SOURCE_EXTENSIONS.contains(&extension.to_lowercase().as_str()))
        .unwrap_or(false)
}

fn relpath(root: &Path, path: &Path) -> String {
    path.strip_prefix(root)
        .unwrap_or(path)
        .to_string_lossy()
        .replace('\\', "/")
}

fn classify_file(path: &Path, rel: &str) -> String {
    let file_name = path.file_name().and_then(|name| name.to_str()).unwrap_or("");
    if CORE_ENTRY_FILES.contains(&file_name) {
        return "entry".to_string();
    }
    let parts: Vec<&str> = rel.split('/').collect();
    let dir_low = parts[..parts.len().saturating_sub(1)].join("/").to_lowercase();
    if !dir_low.is_empty() {
        if contains_any(&dir_low, CORE_TOKENS) {
            return "core".to_string();
        }
        if contains_any(&dir_low, UI_TOKENS) {
            return "ui".to_string();
        }
        if contains_any(&dir_low, SERVICE_TOKENS) {
            return "service".to_string();
        }
    }
    let low = rel.to_lowercase();
    if contains_any(&low, UI_TOKENS) {
        return "ui".to_string();
    }
    if contains_any(&low, SERVICE_TOKENS) {
        return "service".to_string();
    }
    if contains_any(&low, CORE_TOKENS) {
        return "core".to_string();
    }
    "other".to_string()
}

fn contains_any(value: &str, tokens: &[&str]) -> bool {
    tokens.iter().any(|token| value.contains(token))
}

fn extract_imports(path: &Path) -> Vec<String> {
    let text = std::fs::read_to_string(path).unwrap_or_default();
    match path.extension().and_then(|extension| extension.to_str()).unwrap_or("").to_lowercase().as_str() {
        "py" => extract_python_imports(&text),
        "js" | "jsx" | "ts" | "tsx" | "mjs" | "cjs" => extract_js_imports(&text),
        _ => Vec::new(),
    }
}

fn extract_python_imports(text: &str) -> Vec<String> {
    let mut imports = BTreeSet::new();
    let mut ordered = Vec::new();
    for line in text.lines() {
        let trimmed = line.trim_start();
        let candidate = if let Some(rest) = trimmed.strip_prefix("from ") {
            rest.split_whitespace().next()
        } else if let Some(rest) = trimmed.strip_prefix("import ") {
            rest.split_whitespace().next()
        } else {
            None
        };
        if let Some(module) = candidate {
            let module = module.trim_end_matches(',');
            if !module.is_empty() && imports.insert(module.to_string()) {
                ordered.push(module.to_string());
            }
        }
    }
    ordered
}

fn extract_js_imports(text: &str) -> Vec<String> {
    let mut seen = BTreeSet::new();
    let mut ordered = Vec::new();
    for line in text.lines() {
        for quote in ['\'', '"'] {
            let Some(start) = line.find(quote) else { continue };
            let rest = &line[start + 1..];
            let Some(end) = rest.find(quote) else { continue };
            let module = &rest[..end];
            if !module.is_empty()
                && (line.trim_start().starts_with("import ") || line.contains("require("))
                && seen.insert(module.to_string())
            {
                ordered.push(module.to_string());
            }
            break;
        }
    }
    ordered
}
// === ANCHOR: PROJECT_SCAN_END ===
